docs/modules/record-encryption/proc-configuring-record-encryption-filter.adoc

```diff
@@ -35,14 +35,21 @@ filters:
       selector: <KEK_selector_service_name> # <3>
       selectorConfig:
         template: "KEK_${topicName}" # <4>
+      experimental:
+        "encryptionDekRefreshAfterWriteSeconds": 3600 # <6>
+        "encryptionDekExpireAfterWriteSeconds": 7200 # <7>
+        "maxEncryptionsPerDek": 5000000 # <8>
 	 # ...
 ----
 <1> The KMS service name.
 <2> Configuration specific to the KMS provider.
 <3> The Key Encryption Key (KEK) selector to use. The `${topicName}` is a literal understood by the proxy. 
 For example, if using the `TemplateKekSelector` with the template `KEK_$\{topicName}`, create a key for every topic that
 is to be encrypted with the key name matching the topic name, prefixed by the string `KEK_`.
-<4> The template for deriving the KEK, based on a specific topic name. 
+<4> The template for deriving the KEK, based on a specific topic name.
+<5> How long to use a DEK for encryption after it has been created, after this time the Filter will asynchronously obtain a new DEK while continuing using the old one. This can be used to control how quickly we respond to the fact that a KEK has been rotated externally. Note that tuning this value has a cost implication. Shorter durations will result in increased operations on the KMS at encrypt and decrypt time, because more DEKs will be created and used.
```

Comment:
"How long to use a DEK for encryption after it has been created, after this time the Filter will asynchronously obtain a new DEK while continuing using the old one"

I think those words are misleading.  It makes it sound like the cache will refresh the entry immediately.

If I understand the javadoc (`com.github.benmanes.caffeine.cache.Caffeine#refreshAfterWrite(java.time.Duration)`) right,  the refreshment will be initiated on the **next request**.  But that request will see the stale value. 

So, I think this worst case timeline is possible (default KmsCacheConfig values).

```
00m DEK A created 
60m DEK A marked for refresh but refreshment does not take place. DEK A still in the cache.
1h59m request..  caller gets stale DEK A, refreshment initiated.
(DEK b produced)
2h00m request caller gets DEK b
```

Reply:
correct, the refresh is triggered on access

> In contrast to expireAfterWrite, refreshAfterWrite will make a key eligible for refresh after the specified duration, but a refresh will only be actually initiated when the entry is queried.

and stale values are still returned during the refresh op

> The old value (if any) is still returned while the key is being refreshed, in contrast to eviction, which forces retrievals to wait until the value is loaded anew.

Your timeline matches my understanding.

https://github.com/ben-manes/caffeine/wiki/Refresh